Add pipeline runner

cjh1 opened the pull request.

cjh1 commented on oremda/cli/runner.py:
```diff
@@ -0,0 +1,12 @@
+
+@click.command(
+    "runner",
+    short_help="oremda runner",
```
We might want a better name here!

cjh1 replied:
So currently  `oremda run` can be using to run a pipeline from a JSON file. The "runner" is a separate service that communicates with web server to run pipelines executed from the web UI.

cjh1 commented on server/server/main.py:
```diff
@@ -232,3 +142,19 @@ async def create_websocket(
+# Create an endpoint and load it with the methods to expose
+endpoint = WebsocketRPCEndpoint(RpcMethods(), on_connect=[on_connect])
+# add the endpoint to the app
+endpoint.register_route(app, "/rpc")
```
This should be cleaned up, maybe moved to another file ....

cjh1 replied:
I have started doing this, will push soon.

cjh1 replied:
Done

cjh1 commented on oremda/registry.py:
```diff
@@ -125,11 +124,9 @@ def run(
         while len(info.containers) < num_to_run:
             container = None
             try:
-                print(f"On {mpi_rank=}, starting {image_name=}")
```
I didn't mean to remove this, will restore

psavery commented on oremda/pipeline/runner/rpc/client.py:
```diff
@@ -0,0 +1,101 @@
+from oremda.display import NoopDisplayHandle
+
+
+def unique_id():
+    return str(uuid.uuid4())
```
Can we have this in a utils file? Maybe re-use the one in `oremda.utils.id`?

cjh1 replied:
Oops, that was the intention.

psavery commented on oremda/pipeline/runner/rpc/client.py:
```diff
@@ -0,0 +1,101 @@
+    # pipeline.run is a blocking function, run it in a separate thread to free the
+    # server to perform other tasks such as sending notifications
+    # TODO: convert pipeline.run to an async function
+    await asyncio.to_thread(pipeline.run)
```
We can use [this decorator](https://github.com/OpenChemistry/oremda/blob/43951fa176ca8a6be19182c8f2ac12f3809691df/oremda/utils/asyncio.py#L7) on `pipeline.run` (or any blocking function) to make it an async function.

psavery commented on setup.py:
```diff
@@ -32,6 +32,9 @@
         ],
-        "oremda.cli.plugins": ["run = oremda.cli.run:main"],
+        "oremda.cli.plugins": [
+            "run = oremda.cli.run:main",
+            "runner = oremda.cli.runner:main",
```
I didn't realize there is a `run` and a `runner`. What is the difference?

cjh1 replied:
One runs a pipeline on the command line and the new one is a service that interacts with the web server to execute pipelines.

psavery commented on oremda/pipeline/runner/context.py:
```diff
@@ -0,0 +1,72 @@
+)
+
+
+class GlobalContext(BaseModel):
+    plasma_client: PlasmaClient = Field(...)
+    container_client: ContainerClient = Field(...)
+    registry: Registry = Field(...)
+    sessions: Dict[IdType, SessionModel] = {}
+    pipelines: Dict[IdType, PipelineModel] = {}
+    websockets: Dict[IdType, WebsocketModel] = {}
+    session_pipelines: Dict[IdType, Set[IdType]] = {}
+    session_websockets: Dict[IdType, Set[IdType]] = {}
+
+    class Config:
+        arbitrary_types_allowed = True
```
Nice! I like having a global context. This can seriously reduce the number of arguments we have to pass around the program.

Maybe we should eventually start using this in some other places where we are passing around arguments?

alesgenova replied:
The global context here serves the purpose of replacing a future database and it works fine.

But in general I would be strongly against replacing regular function parameters for some global state.

psavery replied:
We have global states in ParaView, Tomviz, Hexrd, etc. They can be very effective if done well. If there is good reason to believe a variable should only have one instance throughout the whole program, then moving it into a global state (such as singleton) can rapidly speed up R&D and the amount of work you can accomplish. The costs of maintaining arguments throughout the program that should be global can be high.

psavery commented on oremda/pipeline/runner/context.py:
```diff
@@ -0,0 +1,72 @@
+    sessions: Dict[IdType, SessionModel] = {}
+    pipelines: Dict[IdType, PipelineModel] = {}
+    websockets: Dict[IdType, WebsocketModel] = {}
+    session_pipelines: Dict[IdType, Set[IdType]] = {}
+    session_websockets: Dict[IdType, Set[IdType]] = {}
```
Just an idea here (that we don't necessarily need to do). What if we had a model like this:
```python
class SessionWebModel(BaseModel):
    session: SessionModel
    pipelines: Set[IdType] = set()
    websockets: Set[IdType] = set()
```

This way we wouldn't need `session_pipelines` or `session_websockets`.

cjh1 replied:
This is copied code, but I can make that improvement.

cjh1 replied:
Done

psavery replied:
This is a good improvement so that we do not have to worry about keeping them synchronized.

psavery commented on oremda/pipeline/runner/.env:
```diff
@@ -0,0 +1,5 @@
+OREMDA_VAR_DIR=/home/cjh/run/oremda
+OREMDA_DATA_DIR=/home/cjh/data/oremda
+OREMDA_DIR=/home/cjh/work/source/oremda
+
```
Should we put the `PLASMA_SOCKET` in here too, as:
```bash
PLASMA_SOCKET=${OREMDA_VAR_DIR}/plasma.sock
```

? I see you can refer to earlier variables like this.

cjh1 replied:
We can do for sure, but as we are adding the `.env` to the `.gitignore` it will be be up to each dev.

cjh1 replied:
Done, I will also remove the file

psavery commented on oremda/pipeline/runner/context.py:
```diff
@@ -0,0 +1,72 @@
+            }
+        )
+
+    registry.release()
```
Should we eventually put this cleanup (`registry.release()`) in a context manager?

cjh1 replied:
Sure, this could be done in a follow up.

psavery commented on oremda/models/__init__.py:
```diff
@@ -0,0 +1,19 @@
+
+
+class ObjectModel(BaseModel):
+    id: IdType = Field(...)
```
Is `id: IdType = Field(...)` preferred over just `id: IdType`? It looks like they are the same: https://pydantic-docs.helpmanual.io/usage/models/#required-fields

cjh1 replied:
The code was just moved into a new file, may be @alesgenova can comment.

alesgenova replied:
Yeah they are the same, I like to put `Field(...)` explicitly so it can be easier to add configuration parameters if needed.

Either way works

psavery reviewed (commented):
Let's meet next week so I can get a high level overview of what is going on...

cjh1 commented on oremda/pipeline/runner/.env:
```diff
@@ -0,0 +1,5 @@
+OREMDA_VAR_DIR=/home/cjh/run/oremda
+OREMDA_DATA_DIR=/home/cjh/data/oremda
+OREMDA_DIR=/home/cjh/work/source/oremda
```
Sure, we can just add it to the `.gitignore`, I just wanted to provide an example.

alesgenova replied:
I think we should keep a somewhat reusable `env.sh` file (usign `$HOME`), even if just to give developers a starting point of the environmental variables that can be set or are expected

cjh1 replied:
You can still use `env.sh`, environment variables will override anything from a dotenv file ...

cjh1 commented:
OK, I have addressed the review comments and done a little more refactoring. The CI is failing because we are importing things from the server which are not install, can we set a PYTHONPATH?

cjh1 commented:
@psavery @alesgenova OK, think I have fixed up things. The last thing is the name. Here are some alternatives to `runner` :- `service`, `pipeline-runner`, `executor`, `engine`. Thoughts?

psavery commented:
> @psavery @alesgenova OK, think I have fixed up things. The last thing is the name. Here are some alternatives to `runner` :- `service`, `pipeline-runner`, `executor`, `engine`. Thoughts?

I like `service`

alesgenova commented:
Either `service` or `engine` sounds good to me

cjh1 commented:
`service` it is then now :smile:

cjh1 commented:
Actually, we will probably end up with `oremda server` to start the fastapi server, so may be `engine` could be better?

psavery commented:
> Actually, we will probably end up with `oremda server` to start the fastapi server, so may be `engine` could be better?

Engine is fine!

psavery reviewed (approved):
LGTM
